aiohttp/payload.py

```diff
@@ -0,0 +1,267 @@
+import asyncio
+import io
+import mimetypes
+import os
+from abc import ABC, abstractmethod
+
+from multidict import CIMultiDict
+
+from . import hdrs
+from .helpers import (content_disposition_header, guess_filename,
+                      parse_mimetype, sentinel)
+from .streams import DEFAULT_LIMIT, DataQueue, EofStream, StreamReader
+
+__all__ = ('PAYLOAD_REGISTRY', 'get_payload', 'Payload',
+           'BytesPayload', 'StringPayload', 'StreamReaderPayload',
+           'IOBasePayload', 'BytesIOPayload', 'BufferedReaderPayload',
+           'TextIOPayload', 'StringIOPayload')
+
+
+class LookupError(Exception):
+    pass
+
+
+def get_payload(data, *args, **kwargs):
+    return PAYLOAD_REGISTRY.get(data, *args, **kwargs)
+
+
+class PayloadRegistry:
+    """Payload registry.
+
+    note: we need zope.interface for more efficient adapter search
+    """
+
+    def __init__(self):
+        self._registry = []
+
+    def get(self, data, *args, **kwargs):
+        if isinstance(data, Payload):
+            return data
+        for ctor, type in self._registry:
+            if isinstance(data, type):
+                return ctor(data, *args, **kwargs)
+
+        raise LookupError()
+
+    def register(self, ctor, type):
+        self._registry.append((ctor, type))
+
+
+class Payload(ABC):
+
+    _size = None
+    _headers = None
+    _content_type = 'application/octet-stream'
+
+    def __init__(self, value, *, headers=None,
+                 content_type=sentinel, filename=None, encoding=None):
+        self._value = value
+        self._encoding = encoding
+        self._filename = filename
+        if headers is not None:
+            self._headers = CIMultiDict(headers)
```

Comment:
Shouldn't here be a multidict proxy? Because, technically, I can update `Payload.headers` content type and it will cause a conflict with the content type property.